index.bs

```diff
@@ -4648,71 +4656,154 @@ The <dfn export for=commands>browsingContext.setViewport</dfn> command modifies
+algorithm is invoked, immediately prior to returning the settings
+object:
+
+1. Let |environment settings| be the newly created [=environment settings
+   object=].
+
+1. Let |related navigables| be the result of [=get related navigables=] given |environment settings|.
+
+1. For each |navigable| of |related navigables|:
+
+   1. If |navigable| is not a [=/top-level traversable=], continue.
+
+   1. Let |user context| be |navigable|'s [=associated user context=].
+
+   1. If [=viewport overrides map=] [=map/contains=] |user context|:
+
+      1. If |viewport overrides map|[|user context|] [=map/contains=] the <code>viewport</code> field:
+
+         1. [=Set viewport=] with |navigable| and |viewport overrides map|[|user context|]["<code>viewport</code>"].
+
+      1. If |viewport overrides map|[|user context|] [=map/contains=] the <code>devicePixelRatio</code> field:
+
+         1. For the |navigable| and all [=descendant navigables=]:
+
+            1. [=Set device pixel ratio override=] with |navigable| and |viewport overrides map|[|user context|]["<code>devicePixelRatio</code>""].
+
 <div algorithm="remote end steps for browsingContext.setViewport">
 
 The [=remote end steps=] with |command parameters| are:
 
-1. Let |navigable id| be the value of the <code>context</code> field of |command
-   parameters|.
+1. If |command parameters| [=map/contains=] "<code>userContexts</code>"
+   and |command parameters| [=map/contains=] "<code>context</code>",
+   return [=error=] with [=error code=] [=invalid argument=].
 
-1. Let |navigable| be the result of [=trying=] to [=get a navigable=] with
-   |navigable id|.
+1. Let |navigables| be a [=/set=].
 
-1. If |navigable| is not a [=/top-level traversable=], return [=error=] with
-   [=error code=] [=invalid argument=].
+1. If the <code>context</code> field of |command parameters| is present:
 
-1. If the implementation is unable to adjust the |navigable|'s [=layout viewport=]
-   parameters with the given |command parameters| for any reason, return
-   [=error=] with [=error code=] [=unsupported operation=].
+   1. Let |navigable id| be the value of the <code>context</code> field of |command
+      parameters|.
 
-1. If |command parameters| [=map/contains=] the <code>viewport</code> field:
+   1. Let |navigable| be the result of [=trying=] to [=get a navigable=] with
+      |navigable id|.
 
-   1. Let |viewport| be the |command parameters|["<code>viewport</code>"].
+   1. If |navigable| is not a [=/top-level traversable=], return [=error=] with
+      [=error code=] [=invalid argument=].
 
-   1. If |viewport| is not null, set the width of |navigable|'s [=layout
-      viewport=] to be the <code>width</code> of |viewport| in CSS pixels and
-      set the height of the |navigable|'s [=layout viewport=] to be the
-      <code>height</code> of |viewport| in CSS pixels.
+   1. If the implementation is unable to adjust the |navigable|'s [=layout viewport=]
```

Comment:
We have this here; what happens if it fails when we try to adjust it in the user context case?